### PR TITLE
Implement checkpoint command

### PR DESCRIPTION
**Pull request by adrianreber**
This enables the checkpoint command. Right now it can be used like this:

```
  # crictl checkpoint --export=/tmp/cp.tar CTR-ID
```

See #662

### CODE REVIEW

**Comment by zouyee** [comment]
Hi! please submit PRs that update **CRI** API  under the [ main repo](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/cri-api)

**Comment by adrianreber** [comment]
Does anyone have any ideas why `critest containerd / main / ubuntu-18.04 / io.containerd.runc.v1 / runc (pull_request)` might have failed. Does not look related to this PR:

```
 Summarizing 2 Failures:
[FAIL][k8s.io] Streaming runtime should support streaming interfaces [It] runtime should support portforward in host network
github.com/kubernetes-sigs/cri-tools/pkg/validate/networking.go:253
[FAIL][k8s.io] Streaming runtime should support streaming interfaces [It] runtime should support portforward [Conformance]
github.com/kubernetes-sigs/cri-tools/pkg/validate/networking.go:253
```

**Comment by saschagrunert** [comment]
@adrianreber this seems to be a flake, let me re-kick the CI

**Comment by adrianreber** [comment]
> @adrianreber this seems to be a flake, let me re-kick the CI

Now it passed. Thanks!

File: cmd/crictl/container.go [hunk 1]
```diff
@@ -757,6 +793,28 @@ func StopContainer(client internalapi.RuntimeService, id string, timeout int64)
...
 }
 
+// CheckpointContainer sends a CheckpointContainerRequest to the server
+func CheckpointContainer(
+	rClient internalapi.RuntimeService,
+	ID string,
+	export string,
+) error {
+	if ID == "" {
+		return fmt.Errorf("ID cannot be empty")
```
Comment (saschagrunert) [inline cmd/crictl/container.go:803]:
```suggestion
		return errors.New("ID cannot be empty")
```

File: go.mod [hunk 2]
```diff
@@ -83,15 +94,15 @@ replace (
...
+	k8s.io/cri-api => k8s.io/kubernetes/staging/src/k8s.io/cri-api v0.0.0-20220714164509-e5f4f8d71b48
 	k8s.io/csi-translation-lib => k8s.io/kubernetes/staging/src/k8s.io/csi-translation-lib v0.0.0-20220503133649-4ce5a8954017
 	k8s.io/kube-aggregator => k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator v0.0.0-20220503133649-4ce5a8954017
 	k8s.io/kube-controller-manager => k8s.io/kubernetes/staging/src/k8s.io/kube-controller-manager v0.0.0-20220503133649-4ce5a8954017
 	k8s.io/kube-proxy => k8s.io/kubernetes/staging/src/k8s.io/kube-proxy v0.0.0-20220503133649-4ce5a8954017
 	k8s.io/kube-scheduler => k8s.io/kubernetes/staging/src/k8s.io/kube-scheduler v0.0.0-20220503133649-4ce5a8954017
 	k8s.io/kubectl => k8s.io/kubernetes/staging/src/k8s.io/kubectl v0.0.0-20220503133649-4ce5a8954017
 	k8s.io/kubelet => k8s.io/kubernetes/staging/src/k8s.io/kubelet v0.0.0-20220503133649-4ce5a8954017
-	k8s.io/kubernetes => k8s.io/kubernetes v1.24.0
+	k8s.io/kubernetes => k8s.io/kubernetes v1.25.0-alpha.2.0.20220714164509-e5f4f8d71b48
```
Comment (saschagrunert) [inline go.mod:105]: Let's move all Kubernetes dependencies to the same commit.

File: cmd/crictl/container.go [hunk 3]
```diff
@@ -588,6 +588,42 @@ var runContainerCommand = &cli.Command{
...
+
+		for i := 0; i < context.NArg(); i++ {
+			containerID := context.Args().Get(i)
+			err := CheckpointContainer(
+				runtimeClient,
+				containerID,
+				context.String("export"),
+			)
+			if err != nil {
+				return fmt.Errorf("Checkpointing the container %q failed: %v", containerID, err)
```
[the remark this answers is not in the document]
Comment (adrianreber) [reply cmd/crictl/container.go]: changed

File: cmd/crictl/container.go (hunk 1 above)
Comment (adrianreber) [reply cmd/crictl/container.go]: changed

File: go.mod (hunk 2 above)
Comment (adrianreber) [reply go.mod]: All Kubernetes dependencies are now pointing to the same commit.

**Comment by adrianreber** [comment]
@saschagrunert Changed as suggested. Can you approve the CI run again? Thanks.

**Review by saschagrunert — APPROVED** [review]

**Review by saschagrunert — CHANGES_REQUESTED** [review]
Oh, the containerd build seems to be broken we now require golang 1.18 for it.

**Comment by adrianreber** [comment]
> Oh, the containerd build seems to be broken we now require golang 1.18 for it.

That seems to be only part of CI which still was using golang 1.17. I updated this 1.18. Let's see if CI likes it better.

@saschagrunert please approve the CI run once more.

**Comment by adrianreber** [comment]
Hmm, seems like cri-tools uses a containerd version from 1.5.x branch. I updated the Github Actions yaml files to point to release 1.6.6, but not sure if this breaks more than it fixes. Let's see.

**Comment by adrianreber** [comment]
CI is happy now.

**Review by saschagrunert — APPROVED** [review]
LGTM, PTAL @feiskyer @mikebrow

**Review by feiskyer — APPROVED** [review]
/lgtm

**Comment by Carol-code2222** [comment]
> This enables the checkpoint command. Right now it can be used like this:
> 
> ```
>   # crictl checkpoint --export=/tmp/cp.tar CTR-ID
> ```
> 
> See #662
```
# crictl restore
```
When will it be implemented and integrated?

**Comment by adrianreber** [comment]
> > This enables the checkpoint command. Right now it can be used like this:
> > ```
> >   # crictl checkpoint --export=/tmp/cp.tar CTR-ID
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > See #662
> 
> ```
> # crictl restore
> ```
> 
> When will it be implemented and integrated?

There is no plan to have a separate `restore` command. You can use `create` and `start` to restore a container. See the test script in CRI-O for an example (https://github.com/cri-o/cri-o/blob/main/test/checkpoint.bats) or https://kubernetes.io/blog/2022/12/05/forensic-container-checkpointing-alpha/#restore-checkpointed-container-standalone

**Comment by Carol-code2222** [comment]
> > > This enables the checkpoint command. Right now it can be used like this:
> > > ```
> > >   # crictl checkpoint --export=/tmp/cp.tar CTR-ID
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > See #662
> > 
> > 
> > ```
> > # crictl restore
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When will it be implemented and integrated?
> 
> There is no plan to have a separate `restore` command. You can use `create` and `start` to restore a container. See the test script in CRI-O for an example (https://github.com/cri-o/cri-o/blob/main/test/checkpoint.bats) or https://kubernetes.io/blog/2022/12/05/forensic-container-checkpointing-alpha/#restore-checkpointed-container-standalone

Is the effect of `create` and `start` exactly the same as that of `restore`? Ensure that the task can be restored to the status at the checkpoint

**Comment by adrianreber** [comment]
> Is the effect of `create` and `start` exactly the same as that of `restore`?

In CRI-O I would say yes. The create code detects if it is a checkpoint image and then tries to prepare everything so that during start the actual restore is done. The `restore` command I used to have was using exactly the same code as going through `create` and `start`.

> Ensure that the task can be restored to the status at the checkpoint

I don't understand this part of the question. But if a checkpoint image is detected only the restore code path is being used.

**Comment by Carol-code2222** [comment]
> > Is the effect of `create` and `start` exactly the same as that of `restore`?
> 
> In CRI-O I would say yes. The create code detects if it is a checkpoint image and then tries to prepare everything so that during start the actual restore is done. The `restore` command I used to have was using exactly the same code as going through `create` and `start`.
> 
> > Ensure that the task can be restored to the status at the checkpoint
> 
> I don't understand this part of the question. But if a checkpoint image is detected only the restore code path is being used.


Has it been tested based on containerd? I'm using containerd, not cri-o. I want to perform c/r based on containerd. If I use `create` and `start` to perform `restore`, is it OK?

**Comment by rst0git** [comment]
@Carol-code2222 CRI-O and containerd currently use different interfaces and checkpoint image format. For example, you can create a checkpoint with containerd and restore it as follows:
```
ctr image pull docker.io/library/redis:alpine
ctr run -d docker.io/library/redis:alpine redis

ctr c checkpoint --rw --task redis checkpoint/redis:1
ctr c restore --rw --live redis-debug checkpoint/redis:1
```

AFAIK the containerd checkpoint/restore functionality is not currently supported via the Kubernetes API. However, there is an open pull request to enable this functionality: https://github.com/containerd/containerd/pull/6965

**Comment by adrianreber** [comment]
What @rst0git said. containerd does not support checkpoint restore via CRI. If you want to use it via CRI (crictl or kubernetes) you can only use CRI-O.

**Comment by Carol-code2222** [comment]
> Kubernetes API

The c/r function of containerd cannot be performed through the Kubernetes API, which sounds bad news for me！！！

**Comment by adrianreber** [comment]
> > Kubernetes API
> 
> The c/r function of containerd cannot be performed through the Kubernetes API, which sounds bad news for me！！！

The PR exists and someone needs to review and merge it. You are welcome to work on this. It works in CRI-O, so you can easily test it. It is still marked as an alpha feature in Kubernetes, so if not all CRI implementations support it there is still time to work on it.

**Comment by Carol-code2222** [comment]
> @Carol-code2222 CRI-O and containerd currently use different interfaces and checkpoint image format. For example, you can create a checkpoint with containerd and restore it as follows:
> 
> ```
> ctr image pull docker.io/library/redis:alpine
> ctr run -d docker.io/library/redis:alpine redis
> 
> ctr c checkpoint --rw --task redis checkpoint/redis:1
> ctr c restore --rw --live redis-debug checkpoint/redis:1
> ```
> 
> AFAIK the containerd checkpoint/restore functionality is not currently supported via the Kubernetes API. However, there is an open pull request to enable this functionality: [containerd/containerd#6965](https://github.com/containerd/containerd/pull/6965)

I run the following `ctr containers` command to try to perform the checkout/restore for the K8S container, but no result is generated and no error is reported.
``` 
ctr -n k8s.io c checkpoint --rw --task 5e7a6e7bc98a77f87f741a7fa34721ed6887825a9cb70c18e5e5516732a456d7 my-nginx
```
At the same time, I tried the following `ctr tasks` command to dump the container process, which was successful, but I don't know how to restore the container process.
```
ctr -n k8s.io tasks checkpoint --exit --image-path=/root/criu-img  --work-path=/root/criu-work  9b90293d96ae0e6aa7435e2732887b2ea4ec9fe615fc999a89b4696a1e14f63a
```
image

**Comment by adrianreber** [comment]
You should ask your questions at containerd and not here. Not sure we can help You here. This PR was about crictl.

**Comment by Carol-code2222** [comment]
> You should ask your questions at containerd and not here. Not sure we can help You here. This PR was about crictl.

I'm sorry, I didn't notice, I just followed the last @rst0git reply and left a message,

**Pull request merged**